types/src/network_address.rs

```diff
@@ -0,0 +1,645 @@
+/// is a valid unicode string. We do this because '/' characters are already our
+/// protocol delimiter and Rust's [`::std::net::ToSocketAddr`] API requires a
+/// `&str`.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DnsName(String);
+
+/// Possible errors when parsing a human-readable [`NetworkAddress`].
+#[derive(Error, Debug)]
+pub enum ParseError {
+    #[error("unknown protocol type: '{0}'")]
+    UnknownProtocolType(String),
+
+    #[error("unexpected end of protocol string")]
+    UnexpectedEnd,
+
+    #[error("error parsing ip4/ip6 address: {0}")]
+    ParseAddrError(#[from] net::AddrParseError),
+
+    #[error("error parsing integer: {0}")]
+    ParseIntError(#[from] num::ParseIntError),
+
+    #[error("error parsing x25519 public key: {0}")]
+    ParseX25519PubkeyError(#[from] CryptoMaterialError),
+
+    #[error("network address cannot be empty")]
+    EmptyProtocolString,
+
+    #[error("protocol string must start with '/'")]
+    InvalidProtocolString,
+
+    #[error("dns name cannot be empty")]
+    EmptyDnsNameString,
+
+    #[error("dns name cannot contain '/' characters")]
+    InvalidDnsNameCharacter,
+
+    #[error("dns name is too long: len: {0} bytes, max len: 255 bytes")]
+    DnsNameTooLong(usize),
+}
+
+#[derive(Error, Debug)]
+#[error("network address cannot be empty")]
+pub struct EmptyError;
+
+///////////////////////
+// RawNetworkAddress //
+///////////////////////
+
+impl RawNetworkAddress {
+    pub fn new(bytes: Vec<u8>) -> Self {
+        Self(bytes)
+    }
+
+    pub fn empty() -> Self {
+        Self(Vec::new())
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+impl Into<Vec<u8>> for RawNetworkAddress {
+    fn into(self) -> Vec<u8> {
+        self.0
+    }
+}
+
+impl AsRef<[u8]> for RawNetworkAddress {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+}
+
+impl TryFrom<&NetworkAddress> for RawNetworkAddress {
+    type Error = lcs::Error;
+
+    fn try_from(value: &NetworkAddress) -> Result<Self, lcs::Error> {
+        let bytes = lcs::to_bytes(value)?;
+        Ok(RawNetworkAddress::new(bytes))
+    }
+}
+
+////////////////////
+// NetworkAddress //
+////////////////////
+
+impl NetworkAddress {
+    fn new(protocols: Vec<Protocol>) -> Self {
+        Self(protocols)
+    }
+}
+
+impl FromStr for NetworkAddress {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s.is_empty() {
+            return Err(ParseError::EmptyProtocolString);
+        }
+
+        let mut protocols = Vec::new();
+        let mut parts = s.split('/');
+
+        if parts.next() != Some("") {
+            return Err(ParseError::InvalidProtocolString);
+        }
+
+        while let Some(protocol_type) = parts.next() {
+            let arg = parts.next().ok_or(ParseError::UnexpectedEnd)?;
```

Comment:
Does the format require that every "protocol" has a "name" + "arg" pair?

Reply:
not required, no. i could go back to parsing each argument depending on the branch, this was just convenient for the current set of supported protocols. what do you think?

Reply:
Why not pass in the parts to the protocol itself? That would be somewhat of the onion that you mentioned before... peel off as much as you need and then return.

Reply:
you mean let the `Protocol` parse from a `&mut Iterator<&str>`? should be possible

Reply:
ok, look at the latest fixup commit. it should be a bit clearer since `Protocol` drives parsing of itself